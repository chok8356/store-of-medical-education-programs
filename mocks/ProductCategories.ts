import { ICategory } from '../types/ICategory'

export const PRODUCTS_CATEGORIES_COUNT = 5

export const categories: ICategory[] = new Array(PRODUCTS_CATEGORIES_COUNT)
  .fill(null)
  .map((x, i) => {
    const id = i + 1
    return {
      id,
      name: `Категория ${id}`
    }
  })
