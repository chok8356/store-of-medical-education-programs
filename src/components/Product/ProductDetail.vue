<template>
  <b-card
    v-if="product"
    class="overflow-hidden">
    <b-row>
      <b-col md="6">
        <b-card-img :src="product.img" />
      </b-col>
      <b-col md="6">
        <b-card-body>
          <h1 class="mb-4">
            {{ product.name }}
          </h1>

          <b-card-text>
            {{ product.description }}
          </b-card-text>

          <b-badge variant="info">
            {{ category.name }}
          </b-badge>

          <div class="d-block mt-4">
            <b-button
              variant="primary"
              @click="addToCart">
              В корзину
            </b-button>
          </div>
        </b-card-body>
      </b-col>
    </b-row>
  </b-card>
</template>

<script lang="ts">
import { BCard, BRow, BCol, BCardBody, BCardText, BBadge, BButton, BCardImg } from 'bootstrap-vue'
import { PropType } from 'vue'
import { Vue, Component, Prop } from 'vue-property-decorator'
import { ICategory } from '../../../types/ICategory'
import { IProduct } from '../../../types/IProduct'

@Component({
  name: 'ProductDetail',
  components: {
    BCard, BRow, BCol, BCardBody, BCardText, BBadge, BButton, BCardImg
  }
})
export default class ProductDetail extends Vue {
  @Prop({ type: Object as PropType<IProduct> }) product!: IProduct
  @Prop({ type: Object as PropType<ICategory> }) category!: ICategory

  addToCart() {
    this.$emit('add-to-cart')
  }
}
</script>
