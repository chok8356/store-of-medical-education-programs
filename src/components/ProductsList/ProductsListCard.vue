<template>
  <b-card
    class="products-list-card"
    :img-src="product.img">
    <b-card-text class="products-list-card__title mb-2">
      {{ product.name }}
    </b-card-text>

    <b-card-text class="font-weight-bold mb-2">
      {{ product.price | currency }}
    </b-card-text>

    <b-card-text class="products-list-card__description">
      {{ product.description | truncate(54) }}
    </b-card-text>

    <b-button-group size="sm">
      <b-button
        variant="outline-primary"
        @click="addToCart">
        В корзину
      </b-button>
      <b-button
        variant="outline-secondary"
        @click="showDetail">
        Подробнее
      </b-button>
    </b-button-group>
  </b-card>
</template>

<script lang="ts">
import { BCard, BCardText, BButtonGroup, BButton } from 'bootstrap-vue'
import { PropType } from 'vue'
import { Vue, Component, Prop } from 'vue-property-decorator'
import { truncate, currency } from '@/filters'
import { IProduct } from '../../../types/IProduct'

@Component({
  name: 'ProductsListCard',
  filters: {
    truncate, currency
  },
  components: {
    BCard, BCardText, BButtonGroup, BButton
  }
})
export default class ProductsListCard extends Vue {
  @Prop({ type: Object as PropType<IProduct> }) product!: IProduct

  addToCart() {
    this.$emit('add-to-cart')
  }

  showDetail() {
    this.$emit('show-detail')
  }
}
</script>

<style lang="scss">
.products-list-card {
  cursor: pointer;

  &__title {
    font-size: 16px;
    font-weight: 500;
  }

  &__description {
    font-size: 13px;
  }
}
</style>
