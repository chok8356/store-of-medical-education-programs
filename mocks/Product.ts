import { IProduct } from '../types/IProduct'
import { PRODUCTS_CATEGORIES_COUNT } from './ProductCategories'

const PRODUCTS_COUNT = 100

export const products: IProduct[] = new Array(PRODUCTS_COUNT).fill(null).map((x, i) => {
  const id = i + 1
  return {
    id,
    name: `Продукт ${id}`,
    price: id + PRODUCTS_COUNT,
    // eslint-disable-next-line max-len
    img: 'https://media.istockphoto.com/photos/stack-of-medical-student-textbooks-picture-id921644532?k=20&m=921644532&s=612x612&w=0&h=hezQpZ0uyUmbU4QJePKK5JPNWVXD_UWwWPYSg5Rspds=',
    // eslint-disable-next-line max-len
    description: id % 7 + ' Полностью новые главы по биохимии и клеточной биологии, генетике, нервной системе, костям, мышцам и коже, эндокринной и репродуктивной системам, сердечно-сосудистой системе, почечной системе, диете и питанию.',
    category: id % PRODUCTS_CATEGORIES_COUNT + 1
  }
})
