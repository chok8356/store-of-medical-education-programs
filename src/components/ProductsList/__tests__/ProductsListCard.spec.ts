import { shallowMount } from '@vue/test-utils'

import { products } from '@/../mocks/Product'
import ProductsListCard from '../ProductsListCard.vue'

const product = products[0]

describe('ProductsListCard', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(ProductsListCard, {
      propsData: {
        product
      }
    })
  })

  afterEach(() => {
    wrapper.destroy()
  })

  it('render correctly', () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it('exist props', () => {
    expect(wrapper.props('product')).toStrictEqual(product)
  })

  it('emit add to cart', () => {
    wrapper.vm.addToCart()
    expect(wrapper.emitted('add-to-cart')).toBeDefined()
  })

  it('emit show detail', () => {
    wrapper.vm.showDetail()
    expect(wrapper.emitted('show-detail')).toBeDefined()
  })
})
