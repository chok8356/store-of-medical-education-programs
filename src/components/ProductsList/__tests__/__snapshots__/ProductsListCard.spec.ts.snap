// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductsListCard render correctly 1`] = `
<b-card-stub
  bodytag="div"
  class="products-list-card"
  footertag="div"
  headertag="div"
  imgsrc="https://media.istockphoto.com/photos/stack-of-medical-student-textbooks-picture-id921644532?k=20&m=921644532&s=612x612&w=0&h=hezQpZ0uyUmbU4QJePKK5JPNWVXD_UWwWPYSg5Rspds="
  subtitletag="h6"
  subtitletextvariant="muted"
  tag="div"
  titletag="h4"
>
  <b-card-text-stub
    class="products-list-card__title mb-2"
    texttag="p"
  >
    
    Продукт 1
  
  </b-card-text-stub>
   
  <b-card-text-stub
    class="font-weight-bold mb-2"
    texttag="p"
  >
    
    101,00 ₽
  
  </b-card-text-stub>
   
  <b-card-text-stub
    class="products-list-card__description"
    texttag="p"
  >
    
    1 Полностью новые главы по биохимии и клеточной биолог...
  
  </b-card-text-stub>
   
  <b-button-group-stub
    ariarole="group"
    size="sm"
    tag="div"
  >
    <b-button-stub
      tag="button"
      target="_self"
      type="button"
      variant="outline-primary"
    >
      
      В корзину
    
    </b-button-stub>
     
    <b-button-stub
      tag="button"
      target="_self"
      type="button"
      variant="outline-secondary"
    >
      
      Подробнее
    
    </b-button-stub>
  </b-button-group-stub>
</b-card-stub>
`;
